Fix reordering product images on multishop

mcdado opened the pull request:
...on a multishop site.

<!-- Thank you for contributing to the PrestaShop project! 

Please take the time to edit the "Answers" rows with the necessary information: -->

| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | 1.6.1.x
| Description?  | When reordering images in a product with the multishop active, when an image is disabled in the current context shop, behind the scenes the `Image` object doesn't load properly, because the query does not return all fields because the LEFT JOIN returns only parts of the fields (see https://cl.ly/0r3c0T1x2d2B/mysql-mistery.mp4 for proof)
| Type?         | bug fix
| Category?     | FO
| BC breaks?    | no
| Deprecations? | no
| Fixed ticket? | n/a
| How to test?  | Enable multishop, add images to a product, disable one of them in a shop and then save the product. Now try to change the order of the images for that product. You might have to look at the browser console for an error message, which is fixed by https://github.com/PrestaShop/PrestaShop/pull/8666/commits/c9e08324c70bed130860ae7d6460ec3943cd82f8 because HTML is being returned to a AJAX request expecting JSON (`SyntaxError: JSON Parse error: Unrecognized token '<'` visible in browser console). The error says `Property Image->id_product is empty`, this error is fixed by https://github.com/PrestaShop/PrestaShop/pull/8666/commits/2051add4f8d64e3c6e2c27936964984d4bf2e8e0. See Description above for details.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/prestashop/prestashop/8666)
<!-- Reviewable:end -->

mcdado committed:
BO: Fix JS error when reordering product images

...on a multishop site.

mcdado commented:
Fixed the PR description. The first commit only fixes the JS error message, I'm working on the root cause.

mcdado committed:
BO: Fix reordering product images in multishop

This happens when some images are not enabled in all shops.

mcdado commented:
With https://github.com/PrestaShop/PrestaShop/pull/8666/commits/82beefd874f6eac850020466ecab9addab901d8a I know I added a method, but I don't see another way of doing this without massive refactorings. If you really don't want another method, it also just be inlined.

mcdado commented:
I amended https://github.com/PrestaShop/PrestaShop/commit/82beefd874f6eac850020466ecab9addab901d8a to avoid adding a method on the Image class, making this change more trivial and with exactly the same results: https://github.com/PrestaShop/PrestaShop/pull/8666/commits/2051add4f8d64e3c6e2c27936964984d4bf2e8e0

TiaNex-Com commented:
yes.there is the problem with reordering the images position when the image doesn't belong to  the default shop,so it can not generate a image object when reordering,

mcdado commented:
Somebody pointed out that the object model hooks are not run this way. Probably the way to do it is to set the missing fields separately.

mcdado committed:
BO: Fix reordering product images

Improves 2051add4f8d64e3c6e2c27936964984d4bf2e8e0 in the sense that
Image->update() is run, thereby all the update hooks are also run.

mcdado committed:
BO: Fix setting cover image for product

This happens when multishop is enabled and the image is not enabled
in the currently selected shop.

mcdado commented:
So now I made sure that all the fields are properly loaded before updating the image.

Also found out that the same problem occurred when setting the cover with an image that is not enabled on the currently selected shop.

mcdado committed:
BO: Fix PHP Notices

mcdado committed:
BO: Fix HTML in AJAX responses after exceptions
